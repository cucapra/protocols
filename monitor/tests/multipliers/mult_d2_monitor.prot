// ARGS: --wave multipliers/mult_d2.fst --instances mult_d2:Multiplier 
struct Multiplier {
  in a: u32,
  in b: u32,
  out s: u32
}

// A multiplier with a 2-cycle delay
fn mul<DUT: Multiplier>(in a: u32, in b: u32, out s: u32) {
  DUT.a := a;
  DUT.b := b;

  step();
  fork();

  DUT.a := X;
  DUT.b := X;

  step();

  DUT.a := X;
  DUT.b := X;

  assert_eq(s, DUT.s);
  step();
}