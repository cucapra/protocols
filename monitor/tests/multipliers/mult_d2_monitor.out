mul(1, 2, 2)
mul(6, 8, 48)
